gen-ai/orchestrator-server/src/main/python/server/src/gen_ai_orchestrator/services/langchain/rag_chain.py

```diff
@@ -238,35 +238,47 @@ def __find_input_variables(template):
+def __rag_guard(inputs, response, documents_required):
     """
-    If a 'no_answer' input was given as a rag setting,
-    then the RAG system should give no further response when no source document has been found.
-    And, when the RAG system responds with the 'no_answer' phrase,
-    then the source documents are removed from the response.
+    Validates the RAG system's response based on the presence or absence of source documents
+    and the `documentsRequired` setting.
+
+    Validation rules:
+    1. When documents_required is True:
+       - The system must only provide answers when relevant documents are found
+       - If no documents are found, the system must either:
+         a) Return the 'no_answer' response if configured
+         b) Raise an exception if 'no_answer' is not configured
+
+    2. When a 'no_answer' response is given:
+       - If documents are present, they are removed and a warning is logged
+       - If no documents are found and documents_required is True, this is considered valid behavior
 
     Args:
         inputs: question answering prompt inputs
         response: the RAG response
+        documents_required (bool): Specifies whether documents are mandatory for the response.
     """
 
-    if 'no_answer' in inputs:
-        if (
-            response['answer'] != inputs['no_answer']
-            and response['source_documents'] == []
-        ):
-            message = 'The RAG gives an answer when no document has been found!'
-            __rag_log(level=ERROR, message=message, inputs=inputs, response=response)
-            raise GenAIGuardCheckException(ErrorInfo(cause=message))
-
-        if (
-            response['answer'] == inputs['no_answer']
-            and response['source_documents'] != []
-        ):
-            message = 'The RAG gives no answer for user question, but some documents has been found!'
-            __rag_log(level=WARNING, message=message, inputs=inputs, response=response)
-            # Remove source documents
-            response['source_documents'] = []
+    no_docs_retrieved = response['source_documents'] == []
+    no_docs_but_required = no_docs_retrieved and documents_required
+    chain_can_give_no_answer_reply = 'no_answer' in inputs
+    chain_reply_no_answer = response.get('answer', '') == inputs.get('no_answer', '')
+
+    if no_docs_but_required:
+        if chain_can_give_no_answer_reply and chain_reply_no_answer:
+            # Expected behavior: chain used 'no_answer' correctly
+            return
+        # Unexpected behavior: chain provided an answer without documents when required
+        message = 'The RAG provides an answer when no document has been found!'
+        __rag_log(level=ERROR, message=message, inputs=inputs, response=response)
+        raise GenAIGuardCheckException(ErrorInfo(cause=message))
+
+    if chain_reply_no_answer and not no_docs_retrieved:
+        # Unexpected behavior: chain replied with 'no_answer' but documents are present
```

Comment:
It's not unexpected sometimes documents found by the retriever doesn't contain the answer to the question. It can happen especially when we don't have any compressor as it will alsways find documents.